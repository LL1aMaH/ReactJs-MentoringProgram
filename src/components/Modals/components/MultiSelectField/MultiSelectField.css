.layout {
  width: 100%;
  position: relative;
}

.dropdown {
  min-width: 100%;
  display: none;
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 3;
  background-color: var(--primary-contrast-color);
  border-radius: 2px;
}

.dropdownOpen {
  display: block;
}

.span {
  margin-top: 2rem;
}
