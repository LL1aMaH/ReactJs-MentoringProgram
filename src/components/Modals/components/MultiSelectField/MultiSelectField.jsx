import React, { memo } from 'react';
import classnames from 'classnames';

import useDropdown from '../../../NavBar/components/SelectField/hooks/useDropdown/useDropdown'

import styles from '../../../../../styles/MultiSelectField.module.scss'

import useMultiSelect from './hooks/useMultiSelect/useMultiSelect';

import { MultiSelectFieldFacade } from './components/MultiSelectFieldFacade/MultiSelectFieldFacade';
import { MultiSelectFieldOptions } from './components/MultiSelectFieldOptions/MultiSelectFieldOptions';


export const MultiSelectField = memo(function MultiSelectField(
  props,
){
  const {
    value,
    onChange,
    options,
    disabled = false,
    placeholder,
    className,
    dropdownClassName,
    label,
    labelClassName,
  } = props;

  const { ref: refDropdown, isOpen, toggleDropdown } = useDropdown();

  const { handleSelect } = useMultiSelect({
    value,
    onChange,
  });

  return (
    <div className={className}>
      <span className={labelClassName}>{label}</span>
      <div ref={refDropdown} className={styles.layout}>
        <MultiSelectFieldFacade onClick={toggleDropdown} active={isOpen}>
          {value.join() ? value.join(' ') : placeholder}
        </MultiSelectFieldFacade>
        <div
          className={classnames(styles.dropdown, dropdownClassName, {
            [styles.dropdownOpen]: isOpen,
          })}
        >
          <MultiSelectFieldOptions
            valueMap={value}
            options={options}
            onChange={handleSelect}
            disabled={disabled}
          />
        </div>
      </div>
    </div>
  );
});
