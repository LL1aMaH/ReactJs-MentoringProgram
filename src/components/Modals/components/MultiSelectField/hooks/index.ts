export { default as useMultiSelect } from './useMultiSelect/useMultiSelect';
