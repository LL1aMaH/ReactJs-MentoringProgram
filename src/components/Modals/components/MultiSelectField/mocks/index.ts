export const mockOptions = [
  'Ann',
  'Jack',
  'Jason',
  'Mary',
  'Carlos',
  'Sara',
  'Matt',
  'Paul',
  'Roberto',
  'Katherine',
  'Robby',
  'Diane',
];
