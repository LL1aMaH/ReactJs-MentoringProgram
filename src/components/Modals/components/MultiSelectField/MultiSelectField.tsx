import React, { memo } from 'react';
import classnames from 'classnames';

import useDropdown from 'Components/NavBar/components/SelectField/hooks/useDropdown/useDropdown';
import { useMultiSelect } from './hooks';

import { MultiSelectFieldFacade } from './components/MultiSelectFieldFacade';
import { MultiSelectFieldOptions } from './components/MultiSelectFieldOptions';

import styles from './MultiSelectField.css';

export type TMultiSelectFieldProps = {
  value: Array<string>;
  onChange: (value: Array<string>) => void;
  options: Array<string>;
  placeholder?: string;
  disabled?: boolean;
  className?: string;
  dropdownClassName?: string;
  label?: string;
  labelClassName?: string;
};

export const MultiSelectField = memo(function MultiSelectField(
  props: TMultiSelectFieldProps,
): JSX.Element {
  const {
    value,
    onChange,
    options,
    disabled = false,
    placeholder,
    className,
    dropdownClassName,
    label,
    labelClassName,
  } = props;

  const { ref: refDropdown, isOpen, toggleDropdown } = useDropdown();

  const { handleSelect } = useMultiSelect({
    value,
    onChange,
  });

  return (
    <div className={className}>
      <span className={labelClassName}>{label}</span>
      <div ref={refDropdown} className={styles.layout}>
        <MultiSelectFieldFacade onClick={toggleDropdown} active={isOpen}>
          {value.join() ? value.join(' ') : placeholder}
        </MultiSelectFieldFacade>
        <div
          className={classnames(styles.dropdown, dropdownClassName, {
            [styles.dropdownOpen]: isOpen,
          })}
        >
          <MultiSelectFieldOptions
            valueMap={value}
            options={options}
            onChange={handleSelect}
            disabled={disabled}
          />
        </div>
      </div>
    </div>
  );
});
