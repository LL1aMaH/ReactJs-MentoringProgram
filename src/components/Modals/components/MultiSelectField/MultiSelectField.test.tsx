import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import { MultiSelectField } from './MultiSelectField';
import { mockOptions } from './mocks';

describe('MultiSelectField', () => {
  const getComponent = (props = {}) =>
    render(
      <MultiSelectField
        value={mockOptions.slice(0, 2)}
        onChange={jest.fn()}
        options={mockOptions}
        placeholder="placeholder"
        {...props}
      />,
    );

  it('should render correctly', () => {
    expect(getComponent({ options: mockOptions }).asFragment()).toMatchSnapshot();
  });

  it('should call onChange with selected values on select option', () => {
    const testOnChange = jest.fn();
    const { getByLabelText } = getComponent({
      value: [mockOptions[3]],
      onChange: testOnChange,
    });

    fireEvent.click(getByLabelText('Ann'));
    expect(testOnChange).toBeCalledWith([mockOptions[3], mockOptions[0]]);
  });

  it('should call onChange with selected values on unselect option', () => {
    const testOnChange = jest.fn();
    const { getByLabelText } = getComponent({
      value: [mockOptions[3], mockOptions[1], mockOptions[10]],
      onChange: testOnChange,
    });

    fireEvent.click(getByLabelText('Jack'));
    expect(testOnChange).toBeCalledWith([mockOptions[3], mockOptions[10]]);
  });
});
