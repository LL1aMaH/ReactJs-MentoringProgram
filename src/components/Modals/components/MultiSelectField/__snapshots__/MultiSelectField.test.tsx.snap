// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectField should render correctly 1`] = `
<DocumentFragment>
  <span
    class="undefined span"
  />
  <div
    class="layout"
  >
    <div
      class="layout"
      data-testid="multi-select-field-facade"
      role="presentation"
    >
      <div
        class="value"
      >
        Ann Jack
      </div>
      <div
        class="icons"
      >
        <svg
          class="icon"
          data-testid="arrow-icon"
          tabindex="-1"
        />
      </div>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="layout"
      >
        <ul
          class="options"
        >
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                checked=""
                class="checkboxField"
                data-testid="checkbox"
                id="Ann"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Ann"
              >
                Ann
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                checked=""
                class="checkboxField"
                data-testid="checkbox"
                id="Jack"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Jack"
              >
                Jack
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Jason"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Jason"
              >
                Jason
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Mary"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Mary"
              >
                Mary
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Carlos"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Carlos"
              >
                Carlos
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Sara"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Sara"
              >
                Sara
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Matt"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Matt"
              >
                Matt
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Paul"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Paul"
              >
                Paul
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Roberto"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Roberto"
              >
                Roberto
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Katherine"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Katherine"
              >
                Katherine
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Robby"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Robby"
              >
                Robby
              </label>
            </div>
          </li>
          <li
            class="option"
          >
            <div
              class="container checkbox"
            >
              <input
                class="checkboxField"
                data-testid="checkbox"
                id="Diane"
                type="checkbox"
                value=""
              />
              <span
                class="checkMark"
              />
              <label
                class="label"
                for="Diane"
              >
                Diane
              </label>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
